tools/site_and_regional/run_neon.py

```diff
@@ -608,6 +608,32 @@ def parse_neon_listing(listing_file, valid_neon_sites):
  
     return available_list
 
+def download_file(url, fname):
```

Comment:
Thanks a lot for your work on this @negin513 and @ekluzek . It's great to have this improved error handling and error messages. I have a few questions:

- [ ] It looks like you added a new function `download_file` but didn't actually replace the old one; I assume you meant to remove the old one below this?

- [ ] [OPTIONAL] If an error is generated – either via a 404 response code or via raising an exception – the script currently prints a message but then continues as if nothing bad had happened. Is that the intended behavior, or should the script either (1) abort or (2) handle this error in some way? I'm not sure what the right answer is here, but just wanted to raise this question.

- [ ] [OPTIONAL] Since the try/except block is really meant to pertain to just the `response = requests.get(url)`, it seems ideal to have just that one line in the try/except. Otherwise, you can inadvertently catch and ignore errors in other code in the `try` block that has nothing to do with the error you're really trying to catch here. This isn't a big deal so I'm marking it as optional, though it seems somewhat more important because you catch all exceptions rather than specific exception types.

Reply:
Thanks for your comment and catching my mistake. 

>  - [ ] It looks like you added a new function `download_file` but didn't actually replace the old one; I assume you meant to remove the old one below this?

1) Yes. `download_file` was duplicated so I remove the duplicated one. I also moved this function to utils.py under python/ctsm as I think this is a function that will be used in other codes. Thanks for your suggestion. This is fixed now. 

> - [ ] [OPTIONAL] If an error is generated – either via a 404 response code or via raising an exception – the script currently prints a message but then continues as if nothing bad had happened. Is that the intended behavior, or should the script either (1) abort or (2) handle this error in some way? I'm not sure what the right answer is here, but just wanted to raise this question. 

I agree with your point 100%. Since this is a generic function that can be used in future, I tried improving upon this. 
I changed the print statements for exception and status_code= 4040 to logger.warnings... Also added I used ctsm.abort to abort the code with error message. Thanks for your suggestion. This is fixed now.

Reply:
Just to make sure I understand this, I tested a few things with a simple program.

For example, If I run the following the code works and raises an exception because `a` is not defined:
```
try: 
    print ("a=", a)
    print ("a is not defined so this try should not work!")
        
    foo = a * 5 
except Exception as err:
    print ("exception is raised because a is not defined!")  
```

Now, if I take `foo=a * 5` outside the try block scope, it gives an error and **the code does NOT work** 
```
try: 
    print ("a=", a)
    print ("a is not defined so this try should not work!")

except Exception as err:
    print ("exception is raised becasue a is not defined!") 

        
foo = a * 5 
```
___________________ 
Similarly I cannot take any lines out of try block in download_file because they are all using `response` object that is defined within the scope of the try block.  Please correct me if I am wrong. I would be happy to hear about this.

Reply:
Thank you for your detailed reply sharing your thoughts on this try/except behavior. You're right that, before you added the abort in the except clause (i.e., my second comment; thank you for doing that!), it would have been problematic to remove the other code from the try block. However, since you are aborting when you catch the exception (which does seem like the right thing to do), you can safely move the rest of the code outside of the try block. The difference in your simple example is that you catch the exception but just carry on instead of aborting: if you aborted in your example (e.g., via a sys.exit call), then you could safely move the setting of `foo = a * 5` outside of the try block. (And, as a general rule, I think that if you catch an exception and don't abort, then you should do something to recover from the exception so that the following code works correctly. So, in this example, you might do something like setting `a` to a default value so that further code can work reasonably.)

With the current code, it doesn't matter hugely. But this could become more important if more code is added to this function in the future – basically, if any code is added that could raise different types of exceptions that arise for different reasons.

Reply:
Aha! Thanks for clarifications! You are right. If we have a sys.exit() or ctsm.abort() to abort the code while the try fails it does not hit the lines outside the try block. But without the sys.exit() or ctsm.abort(), it would not have worked. 
So to summarize this works and is preferable:

```python
try: 
    print ("a=", a)
    print ("a is not defined so this try should not work!")
    sys.exit()                                                                                                                                 
        
except Exception as err:
    print ("exception is raised becasue a is not defined!") 
        
        
foo = a * 5 
```
compared to 
```python
try: 
    print ("a=", a)
    print ("a is not defined so this try should not work!")
    sys.exit()
    foo = a * 5                                                                                                                                  
        
except Exception as err:
    print ("exception is raised becasue a is not defined!") 
```        
        



Since I implemented your second comment and added abort, I moved the response lines outside the try block and implemented all your comments.

Reply:
GTG